returnn/util/file_cache.py

```diff
@@ -114,6 +119,8 @@ def get_file(self, src_filename: str) -> str:
             try:
                 self._copy_file_if_needed(src_filename, dst_filename)
                 break
+            except FileNotFoundError as e:
```

Comment:
I don't understand why you catch this. This sounds like an exception which should immediately lead to an error. We should not repeat when this happens.

Reply:
See #1550. This error occurs when the target directory is removed. If the source file is missing, the retries don't hurt (and will simply delay the process for 3s, but 🤷🏼‍♂️).

Reply:
Ah, but then I think I would make this more explicit. Check for this exception in `_copy_with_prealloc` when opening the `dst` file, and if you get it there, rethrow a custom exception (`_CannotOpenDirWasRemovedException`or whatever), and catch this here.

But despite this, I wonder a bit how this could even happen. The empty dir cleanup should not clean dirs which were recently created.

Reply:
I think I would not do this logic at all currently. To me it sounds like there was a different bug that the directory was cleaned (removed) although it should not have been removed. If this is done correctly, we never should run into the problem here, thus I would also not add this logic, but rather fix the bug itself.

The question is, why was it removed. I think we did not properly update its time. Why? I think the cache dir lock was not doing this anymore. Before, via `with self._touch_files_thread.files_added_context([dst_dir, lock.lockfile]):`, we kept this dir alive and updated.

Maybe also the check `cur_time - os.stat(root).st_mtime <= cur_used_time_threshold` when cleaning up empty dirs is really a bit too aggressive.

Reply:
I'm actually not a big fan of removing the retry logic. I get your point that we should look at the origin of the issue, but adding multiple layers of resilience to a piece of code that is inherently very prune to race conditions and messy behavior does not hurt IMO.

Reply:
There are multiple problems with it:

* It adds further complexity.
* It hides other bugs and makes debugging harder.

Without really having any benefit. I don't see how it can possibly get this wrong when there is no bug in our code. I don't like to add workarounds for bugs in our code just because we don't want to fix the bugs. After we fixed the bug, this really should never ever be necessary. There is no race condition here.

And the bug fix is probably very simple also. As said, I think already by doing the lock again on file level, the `files_added_context` on the lock file should handle this already (it will also update the dirs, see `_TouchFilesThread`).

Reply:
Then we should be good now. The lock is back at file level after this PR.

Reply:
Yes exactly. Unless we miss something. But then there is a bug we should fix. Because I think we really should not run into this problem otherwise.

Maybe just to be a bit saver, we should increase the threshold in `cur_time - os.stat(root).st_mtime <= cur_used_time_threshold` when cleaning the dirs.

Reply:
So, choosing between (1) some more crashes now but overall more stable code after solving this vs (2) less crashes now but maybe some bugs hidden and we forget about it, I would choose (1). (2) leads to the situation that RETURNN is buggier over time, and this over medium/long term is even more annoying to people.